### DISCUSSION TITLE
Config option naming

### DISCUSSION THREAD

**Discussion by schmidtjonathan** [probabilistic-numerics/probnum#497] [General]
In the context of PR #479 , a new configuration option `prefer_dense_arrays` is established, which controls whether or not `LinearOperator`s are used (wherever possible) instead of dense arrays, in order to speed up (lazify) linalg-based computations.
For now, the option is only used in the `statespace.IBM` (integrated Brownian motion process), but its use will be extended rather sooner than later.
Since this option is one that might be relevant beyond the scope of `statespace`, `filtsmooth`, etc. , I wanted to ask for your opinions regarding naming and default value.
@pnkraemer and me would propose:

`name = prefer_dense_arrays`
`default = True`

If you have an opinion on this, feel free to propose changes. Otherwise we would merge this in the following couple of days.

**Comment by schmidtjonathan** [comment]
@JonathanWenger @marvinpfoertner @nathanaelbosch FYI

**Reply by JonathanWenger** [reply]
How about just 'lazy' or a variant including linear algebra / operators?

**Reply by pnkraemer** [reply]
Thanks for replying! :)

I am not sure about just `lazy`, because it might not be super clear what this actually stands for. What I like about `prefer_dense_arrays` (or even `prefer_dense_array_implementation`) is that it is super verbose and there is no doubt about what it means. Or is it just me that thinks this way?

The idea behind `dense_array` instead of `linops` (e.g. `use_linops`) was kind of twofold:
1. I am not sure whether in the near(ish) to mid-term future, we will not wrap most system matrices through linops (i.e. by having a `DenseMatrixLinop` (naming irrelevant here). In this case, a `use_linops` would not be very informative anymore?! On the other hand, `use_sparse_linops` might not always be meaningful as well, e.g. in the sense that Kronecker matrices are not really sparse, are they?
2. Making the user explicitly write code that essentially says "I want a dense array, yes please, I know what this might slow down"  suggests a nice default mindset :) But then again, this might be personal views...

Something that might combine your suggestions and our reasons above would be `lazy_linops`, but I am not sure about this either...

**Reply by JonathanWenger** [reply]
The reason I was hesitant about `prefer_dense_arrays` is that to me it was not quite clear what this would mean. In particular, does this mean sparse arrays are converted to `np.ndarray`s? Something you would essentially never want as a config, I think. It also masks the fact that in-place operations might be preferrable in some instances. I agree that more verbosity than `lazy` is better and `lazy_linops` is better I think. I would also suggest:

```python
lazy_matrix_multiplication
lazy_linalg
memory_efficient_linalg
```
In particular i prefer `lazy_*` and `memory_efficient_*` since they are more descriptive in what the intention and effect is than with `dense_arrays`.

As a side note, `DenseMatrixLinop` essentially exists as `linops.Matrix`, which is a wrapper for `np.ndarray`s and `scipy.sparse.spmatrix`.

**Reply by pnkraemer** [reply]
Okay, if `prefer_dense_arrays` was not clear then this is something to take into account. Thanks for mentioning that.

In this case I would probably opt for `lazy_linops` (which - on a side note - is also a fun alliteration! :smile:)

* `lazy_matrix_multiplication` sounds like it does not apply to sums of linear operators
* `lazy_linalg` and `memory_efficient_linalg` sound a bit like they only affect `probnum.linalg`. This would be solved if we replaced `linalg` with `arithmetic`, i.e. `lazy_arithmetic` or `memory_efficient_arithmetic`, but I am not sure whether I find those better than `lazy_linops`.

**Reply by JonathanWenger** [reply]
So to me there are two parts with linear operators:
- their arithmetic is lazy
- they provide memory-efficient matrix-vector products

In that way `lazy_linops` is somewhat of a misnomer, since it is not the linear operator itself that is lazy (its action is computed immediately if you execute `A @ x`, but rather the arithmetic. 

How about one of the following options:
- `prefer_linops`
- `lazify_with_linops`
and clearly write into the config docstring the properties of linear operators.

**Reply by schmidtjonathan** [reply]
I also like `lazy_linalg`, for the alliteration alone. 😄 Even though, I get your point, @pnkraemer , I also prefer `lazy_linalg` over `lazy_linops`.

`prefer_linops` is also nice, but less expressive, probably.

**Reply by schmidtjonathan** [reply]
How about we vote? (You should be able to edit my comment, just add your initials after the ":")

* `prefer_linops` : 
* `lazy_linalg` :  JS 
* `lazify_with_linops` : JS
* `matrix_free_linalg` : JW, MP

**Reply by marvinpfoertner** [reply]
How do you guys feel about `matrix_free_linalg`? I think this describes exactly what we want to say while also using established terminology

**Reply by pnkraemer** [reply]
I see all your points. I would say my concern (that `*_linalg` sounds like it belongs only to `probnum.linalg`) is weaker than the concern with `*_linops` (because this does not really fit the scope of what is happening -- linear operators may be more general than that`) or `*_arithmetic` (I see that this might not 100% capture linop-vector interaction). 

In light of this, I am fine with both `lazy_linalg`, and `matrix_free_linalg`.
